akka-distributed-data/src/main/scala/akka/cluster/ddata/ORMap.scala

```diff
@@ -125,11 +214,42 @@ final class ORMap[A, B <: ReplicatedData] private[akka] (
    * INTERNAL API
    */
   @InternalApi private[akka] def updated(node: UniqueAddress, key: A, initial: B)(modify: B ⇒ B): ORMap[A, B] = {
-    val newValue = values.get(key) match {
-      case Some(old) ⇒ modify(old)
-      case _         ⇒ modify(initial)
+    val (oldValue, hasOldValue) = values.get(key) match {
+      case Some(old) ⇒ (old, true)
+      case _         ⇒ (initial, false)
     }
-    new ORMap(keys.add(node, key), values.updated(key, newValue))
+    // Optimization: for types - like GSet, GCounter, PNCounter and ORSet  - that are delta based
+    // we can emit (and later merge) their deltas instead of full updates
+    // for the cases where there is no previous value, the delta will not be different from full value
+    // therefore we emit full value via PutDeltaOp instead
+    def modifyWithDelta(value: DeltaReplicatedData) = {
+      val newValue = modify(value.resetDelta.asInstanceOf[B])
+      val newValueDelta = newValue.asInstanceOf[DeltaReplicatedData].delta
+      (newValue, newValueDelta)
+    }
+
+    // propagate 'put' where there was no old value - so effectively we put a new value
+    def gatherDelta(newValue: B, newValueDelta: Option[ReplicatedDelta]) = if (newValueDelta.isDefined && hasOldValue)
+      (newValue, UpdateDeltaOp(keys.resetDelta.add(node, key).delta.get, Map(key → newValueDelta.get), node))
+    else
+      (newValue, PutDeltaOp(keys.resetDelta.add(node, key).delta.get, Map(key → newValue), node))
+
+    val (newValue, modifiedDelta) = oldValue match {
+      case _: DeltaReplicatedData ⇒
+        val (newValue, newValueDelta) = modifyWithDelta(oldValue.asInstanceOf[DeltaReplicatedData])
+        gatherDelta(newValue, newValueDelta)
+      case _ ⇒
+        val newValue = modify(oldValue)
+        (newValue, PutDeltaOp(keys.resetDelta.add(node, key).delta.get, Map(key → newValue), node))
+    }
+    val newDelta = delta match {
+      case Some(d) ⇒
+        d.merge(modifiedDelta)
+      case None ⇒
+        modifiedDelta
+    }
+    // if updated is addition of previously removed key we need to remove the tombstone
+    new ORMap(keys.add(node, key), values.updated(key, newValue), tombstoneVals - key, Some(newDelta))
```

Comment:
After `update` the value is now surely whole - we will prefer it to whatever comes from other node. So no need to keep previous state in tombstone. Hence `tombstoneVals - key`